### PR TITLE
[AIRFLOW-3391] Upgrade pendulum to latest major version

### PR DESCRIPTION
**Pull request by crhyatt**
There is a bug in the current dependency version of pendulum that Airflow which is fixed after pendulum version 2.0.0.  [AIRFLOW-3391] updates pendulum to version ~2.0.

---
Issue link: [AIRFLOW-3391](https://issues.apache.org/jira/browse/AIRFLOW-3391)

Make sure to mark the boxes below before creating PR: [x]

- [X] Description above provides context of the change
- [X] Commit message/PR title starts with `[AIRFLOW-NNNN]`. AIRFLOW-NNNN = JIRA ID<sup>*</sup>
- [X] Unit tests coverage for changes (not needed for documentation changes)
- [X] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [X] Relevant documentation is updated including usage instructions.
- [X] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

<sup>*</sup> For document-only changes commit message can start with `[AIRFLOW-XXXX]`.

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

**Comment by dimberman** [comment]
Hi @crhyatt could you please rebase and ping me? Will gladly review when it's up and running :)

**Comment by crhyatt** [comment]
Hi @dimberman.  I have rebased the PR.  The new Pendulum packages has some slight differences.  Let me know if I need make some changes.

**Comment by crhyatt** [comment]
@dimberman Can you explain the Codecov report to me?

**Comment by dimberman** [comment]
@crhyatt I have no idea why it's doing that... can you rebase again? I'm not too worried about codecov here. @kaxil maybe you've seent this before?

**Comment by kaxil** [comment]
> @crhyatt I have no idea why it's doing that... can you rebase again? I'm not too worried about codecov here. @kaxil maybe you've seent this before?

CodeCov basically tells you what part of the code is not covered by tests but ignore it as it has been unstable off-late.

File: airflow/models/dag.py [hunk 1]
```diff
@@ -411,15 +411,15 @@ def following_schedule(self, dttm):
             if not self.is_fixed_time_schedule():
                 # relative offset (eg. every 5 minutes)
                 delta = cron.get_next(datetime) - naive
-                following = dttm.in_timezone(self.timezone).add_timedelta(delta)
+                following = dttm.in_timezone(self.timezone).__add__(delta)
```
Comment (ashb) [inline airflow/models/dag.py]: Don't do this -- use of `__add__` or any dunder (double underscore) except `__class__` is a huge code smell.

```suggestion
                following = dttm.in_timezone(self.timezone) + delta
```

File: airflow/models/dag.py [hunk 2]
```diff
@@ -1669,7 +1669,7 @@ def get_serialized_fields(cls):
             cls.__serialized_fields = frozenset(vars(DAG(dag_id='test')).keys()) - {
                 'parent_dag', '_old_context_manager_dags', 'safe_dag_id', 'last_loaded',
                 '_full_filepath', 'user_defined_filters', 'user_defined_macros',
-                'partial', '_old_context_manager_dags',
+                '_schedule_interval', 'partial', '_old_context_manager_dags',
```
Comment (ashb) [inline airflow/models/dag.py]: Why did you add this back?

File: setup.py [hunk 3]
```diff
@@ -578,7 +578,7 @@ def write_version(filename: str = os.path.join(*[my_dir, "airflow", "git_version
     'lockfile>=0.12.2',
     'markdown>=2.5.2, <3.0',
     'pandas>=0.17.1, <2.0',
-    'pendulum==1.4.4',
+    'pendulum==2.0.5',
```
Comment (ashb) [inline setup.py]:
```suggestion
    'pendulum~=2.0',
```

**Review by ashb — CHANGES_REQUESTED** [review]
Haven't yet given this a full review, but some changed needed.

Additionally Airflwo-5575 is not the right Jira to target for upgrading pendulum -- it should be AIRFLOW-3391

File: airflow/utils/timezone.py [hunk 4]
```diff
@@ -176,4 +176,4 @@ def parse(string, timezone=None):
 
     :param string: time string
     """
-    return pendulum.parse(string, tz=timezone or TIMEZONE)
+    return pendulum.parse(string, tz=timezone or TIMEZONE, strict=False)
```
Comment (ashb) [inline airflow/utils/timezone.py:179]: :interrobang:

File: requirements/requirements-python3.7.txt [hunk 5]
```diff
@@ -291,7 +291,7 @@ rsa==4.0
 s3transfer==0.3.3
 sasl==0.2.1
 semver==2.9.1
-sendgrid==6.2.1
+sendgrid==6.2.2
```
Comment (ashb) [inline requirements/requirements-python3.7.txt]: @potiuk Is it expected that we see changes like this from non-master/cron changes? It's not the end of the world, but it makes it harder to see what caused a version to get upgraded if random PRs update these deps.

File: tests/models/test_taskinstance.py [hunk 6]
```diff
@@ -589,7 +589,7 @@ def test_next_retry_datetime_short_intervals(self):
 
         date = ti.next_retry_datetime()
         # between 1 * 2^0.5 and 1 * 2^1 (15 and 30)
-        period = ti.end_date.add(seconds=1) - ti.end_date.add(seconds=15)
+        period = pendulum.period(ti.end_date.add(seconds=1), ti.end_date.add(seconds=15))
```
Comment (ashb) [inline tests/models/test_taskinstance.py]: This change shouldn't be needed, right? Cos dt - dt should equal a period?

File: airflow/utils/timezone.py (hunk 4 above)
Comment (crhyatt) [reply airflow/utils/timezone.py:179]: The new version of pendulum defaults strict=True for the parse function.  Which does not allow the parser to fall back to the dateutil parser.  [https://pendulum.eustace.io/docs/#parsing]

When the test tests/www/test_views.py::TestLogView::test_get_file_task_log_0 passes the pendulum parser a datetime default formatted string, like 2017-09-01 00:00:00+00:00.  Which is not a strict ISO8601 or RFC3339 format.

File: setup.py (hunk 3 above)
Comment (crhyatt) [reply setup.py]: ok

File: airflow/models/dag.py (hunk 2 above)
Comment (crhyatt) [reply airflow/models/dag.py]: I think this was a bad rebase/merge on my part

**Comment by crhyatt** [comment]
@Ashb The PR for AIRFLOW-3391 has been created.

**Review by crhyatt — COMMENTED** [review]
@ashb would you like me to cancel this PR and move to 3391?

**Comment by ashb** [comment]
> @ashb would you like me to cancel this PR and move to 3391?

Just rename this PR and update the commit

File: tests/www/test_views.py [hunk 7]
```diff
@@ -883,7 +882,7 @@ def test_run_with_not_runnable_states(self, get_default_executor_function):
             self.check_content_in_response('', resp, resp_code=200)
 
             msg = "Task is in the &#39;{}&#39; state which is not a valid state for execution. " \
-                  .format(state) + "The task must be cleared in order to be run"
+                .format(state) + "The task must be cleared in order to be run"
```
Comment (ashb) [inline tests/www/test_views.py:958]: You have a lot of whitespace-only/formatting-only changes here.

Please remove them from this PR (having unrelated changes in like this in a PR makes it harder to back-port fixes.)

**Review by ashb — CHANGES_REQUESTED** [review]

**Comment by crhyatt** [comment]
@ashb I messed something up with a merge.  I am going to close this PR and just use https://github.com/apache/airflow/pull/9184 .

**Pull request closed without merging**